### PR TITLE
UCM/TEST-EVENTS: added test_external_events call

### PR DESCRIPTION
**Pull request by hoopoepg**
- added ucm_test_external_events function
- added tests for new function
- bit update test for bistro hooks

### CODE REVIEW

File: src/ucm/api/ucm.h [hunk 1]
```diff
@@ -323,6 +323,21 @@ void ucm_unset_external_event(int events);
 ucs_status_t ucm_test_events(int events);
 
 
+/**
+ * @brief Test event external handlers
+ *
+ * This routine checks if event handlers are called when corresponding system API
+ * is invoked through external infrastructure, see @ref ucm_set_external_event
+ * for details.
```
Comment (yosefe) [inline src/ucm/api/ucm.h:331]: `This routine checks if external events, as set by @ref ucm_set_external_event, are actually being reported (by calling APIs such as @ref ucm_vm_munmap)`

File: src/ucm/mmap/install.c [hunk 2]
```diff
@@ -24,6 +24,7 @@
 #include <ucs/sys/preprocessor.h>
 #include <ucs/arch/bitops.h>
 #include <ucs/debug/assert.h>
+#include <ucs/debug/log.h>
```
Comment (yosefe) [inline src/ucm/mmap/install.c:27]: must use UCM log, not UCS log

File: src/ucm/mmap/install.c [hunk 3]
```diff
@@ -90,6 +98,25 @@ static ucm_mmap_func_t ucm_mmap_funcs[] = {
...
+    UCM_MMAP_EVENT_NAME_ENTRY(UCM_EVENT_SBRK),
+    UCM_MMAP_EVENT_NAME_ENTRY(UCM_EVENT_MADVISE),
+
+    /* Aggregate events */
+    UCM_MMAP_EVENT_NAME_ENTRY(UCM_EVENT_VM_MAPPED),
+    UCM_MMAP_EVENT_NAME_ENTRY(UCM_EVENT_VM_UNMAPPED),
+
+    /* Non-accessible memory alloc/free events */
+    UCM_MMAP_EVENT_NAME_ENTRY(UCM_EVENT_MEM_TYPE_ALLOC),
+    UCM_MMAP_EVENT_NAME_ENTRY(UCM_EVENT_MEM_TYPE_FREE)
```
Comment (yosefe) [inline src/ucm/mmap/install.c:117]: why need entry for this?

File: src/ucm/mmap/install.c [hunk 4]
```diff
@@ -50,6 +51,13 @@
         (_data)->out_events &= ~exp_events | (_data)->fired_events;           \
     } while(0)
 
+#define UCM_MMAP_EVENT_NAME_ENTRY(_event) \
+    [ucs_ilog2(_event)] = #_event
```
Comment (yosefe) [inline src/ucm/mmap/install.c:55]: maybe use shorter name like "mmap", "munmap", etc

File: src/ucm/mmap/install.c [hunk 5]
```diff
@@ -50,6 +51,13 @@
...
+
+#define UCM_MMAP_REPORT_BUF_LEN /* use longest event name as base */ \
+    ((sizeof("UCM_EVENT_MEM_TYPE_ALLOC") + 2) * \
```
Comment (yosefe) [inline src/ucm/mmap/install.c:58]: looks weird, define max event name len instead

File: src/ucm/mmap/install.c [hunk 6]
```diff
@@ -194,6 +221,36 @@ void ucm_fire_mmap_events(int events)
...
+    buf = buf_p = malloc(UCM_MMAP_REPORT_BUF_LEN);
+    if (buf == NULL) {
+        ucm_error("Out of memory");
+        return;
+    }
+
+    end_p = buf_p + UCM_MMAP_REPORT_BUF_LEN;
+    ucs_for_each_bit(idx, expected) {
+        if (!(actual & UCS_BIT(idx))) {
+            snprintf(buf_p, end_p - buf_p, "%s%s", (events_count++ ? ", " : ""),
```
Comment (yosefe) [inline src/ucm/mmap/install.c:241]: do ++ outside of func args

File: src/ucm/mmap/install.c [hunk 7]
```diff
@@ -194,6 +221,36 @@ void ucm_fire_mmap_events(int events)
...
+    char *end_p;
+
+    buf = buf_p = malloc(UCM_MMAP_REPORT_BUF_LEN);
+    if (buf == NULL) {
+        ucm_error("Out of memory");
+        return;
+    }
+
+    end_p = buf_p + UCM_MMAP_REPORT_BUF_LEN;
+    ucs_for_each_bit(idx, expected) {
```
Comment (yosefe) [inline src/ucm/mmap/install.c:239]: ucs_for_each_bit(idx, expected & ~actual)

File: src/ucm/mmap/install.c [hunk 8]
```diff
@@ -194,6 +221,36 @@ void ucm_fire_mmap_events(int events)
...
+        if (!(actual & UCS_BIT(idx))) {
+            snprintf(buf_p, end_p - buf_p, "%s%s", (events_count++ ? ", " : ""),
+                     ucm_mmap_event_name[idx]);
+            buf_p += strlen(buf_p);
+        }
+    }
+
+    if (events_count) {
+        ucm_diag("Missing events: %s", buf);
```
Comment (yosefe) [inline src/ucm/mmap/install.c:248]: ucm_diag("missing %s memory events: %s", title, buf);
where title is arg for ucm_mmap_event_report_missing, can be "internal" or "external"

File: src/ucm/mmap/install.c [hunk 9]
```diff
@@ -194,6 +221,36 @@ void ucm_fire_mmap_events(int events)
...
 
+static void ucm_mmap_event_report_missing(int expected, int actual)
+{
+    int events_count = 0;
+    int idx;
+    char *buf;
+    char *buf_p;
+    char *end_p;
+
+    buf = buf_p = malloc(UCM_MMAP_REPORT_BUF_LEN);
```
Comment (yosefe) [inline src/ucm/mmap/install.c:232]: maybe allocate on stack, it's not supposed to be large

File: src/ucm/mmap/install.c [hunk 10]
```diff
@@ -215,27 +272,36 @@ static ucs_status_t ucm_mmap_test_events(int events)
...
     if (!ucs_test_all_flags(data.out_events, events)) {
+        ucm_mmap_event_report_missing(events, data.out_events);
         return UCS_ERR_UNSUPPORTED;
     }
 
     return UCS_OK;
 }
 
-ucs_status_t ucm_mmap_test_installed_events(int events)
+ucs_status_t ucm_mmap_test_events_safe(int events)
```
Comment (yosefe) [inline src/ucm/mmap/install.c:282]: ucm_mmap_test_events_safe -> ucm_mmap_test_events
ucm_mmap_test_events -> ucm_mmap_test_events_nolock

Comment (hoopoepg) [reply src/ucm/mmap/install.c]: renamed

**Comment by hoopoepg** [comment]
@yosefe ok to squash?

File: src/ucm/mmap/install.c [hunk 11]
```diff
@@ -231,11 +231,10 @@ static void ucm_mmap_event_report_missing(int expected, int actual,
 
     buf   = buf_p = ucs_alloca(UCM_MMAP_REPORT_BUF_LEN);
     end_p = buf_p + UCM_MMAP_REPORT_BUF_LEN;
-    ucs_for_each_bit(idx, expected & ~actual) {
-        if (idx > ucs_log2(UCM_EVENT_VM_UNMAPPED)) { /* for coverity */
-            break;
-        }
 
+    ucs_for_each_bit(idx,
+                     expected & ~actual & ((UCM_EVENT_VM_UNMAPPED << 1) - 1)) {
```
Comment (yosefe) [inline src/ucm/mmap/install.c:236]: maybe 
int missing_events;
missing_events = expected & ~actual & UCS_MASK(UCM_EVENT_VM_UNMAPPED)
BTW it should be `1<< UCM_EVENT_VM_UNMAPPED`, not `UCM_EVENT_VM_UNMAPPED << 1`

File: src/ucm/event/event.c [hunk 12]
```diff
@@ -621,6 +621,11 @@ ucs_status_t ucm_test_events(int events)
     return ucm_mmap_test_installed_events(events);
 }
 
+ucs_status_t ucm_test_external_events(int events)
+{
+    return ucm_mmap_test_events(events & ucm_external_events, 1);
```
Comment (yosefe) [inline src/ucm/event/event.c:626]: can pass the string "external" to ucm_mmap_test_events() instead of 1

File: src/ucm/mmap/install.c [hunk 13]
```diff
@@ -194,8 +218,37 @@ void ucm_fire_mmap_events(int events)
...
+
+    UCS_STATIC_ASSERT(UCM_MMAP_REPORT_BUF_LEN <= UCS_ALLOCA_MAX_SIZE)
+
+    buf   = buf_p = ucs_alloca(UCM_MMAP_REPORT_BUF_LEN);
+    end_p = buf_p + UCM_MMAP_REPORT_BUF_LEN;
+
+    ucs_for_each_bit(idx,
+                     expected & ~actual & ((UCM_EVENT_VM_UNMAPPED << 1) - 1)) {
+        /* coverity[overrun-local] */
+        snprintf(buf_p, end_p - buf_p, "%s%s", (events_count ? ", " : ""),
```
Comment (yosefe) [inline src/ucm/mmap/install.c:238]: `(events_count > 0) ? ", " : ""`

File: src/ucm/event/event.c (hunk 12 above)
Comment (hoopoepg) [reply src/ucm/event/event.c]: done

File: src/ucm/mmap/install.c (hunk 11 above)
Comment (hoopoepg) [reply src/ucm/mmap/install.c]: added variable

> BTW it should be `1<< UCM_EVENT_VM_UNMAPPED`, not `UCM_EVENT_VM_UNMAPPED << 1`

nop, ```UCM_EVENT_VM_UNMAPPED == UCS_BIT(17)```, and ```1<< UCM_EVENT_VM_UNMAPPED``` equals ```1 << UCS_BIT(17)``` - it is out of ```uint64_t```

File: src/ucm/mmap/install.c (hunk 13 above)
Comment (hoopoepg) [reply src/ucm/mmap/install.c]: hmmm, 0 - is initial value, non-zero - all other states
but ok, fixed

File: src/ucm/mmap/install.c [hunk 14]
```diff
@@ -219,36 +219,37 @@ void ucm_fire_mmap_events(int events)
...
     char *buf_p;
     char *end_p;
 
     UCS_STATIC_ASSERT(UCM_MMAP_REPORT_BUF_LEN <= UCS_ALLOCA_MAX_SIZE)
 
-    buf   = buf_p = ucs_alloca(UCM_MMAP_REPORT_BUF_LEN);
-    end_p = buf_p + UCM_MMAP_REPORT_BUF_LEN;
+    buf            = buf_p = ucs_alloca(UCM_MMAP_REPORT_BUF_LEN);
+    end_p          = buf_p + UCM_MMAP_REPORT_BUF_LEN;
+    missing_events = expected & ~actual & ((UCM_EVENT_VM_UNMAPPED << 1) - 1);
```
Comment (yosefe) [inline src/ucm/mmap/install.c:235]: maybe use UCS_MASK(ucs_array_size(ucm_mmap_event_name))
instead of
((UCM_EVENT_VM_UNMAPPED << 1) - 1);

Comment (hoopoepg) [reply src/ucm/mmap/install.c]: good idea :)
done

**Pull request merged**